tools/kagami/src/crypto.rs

```diff
@@ -1,14 +1,135 @@
+    public_key_string: Option<String>,
+    /// Path to public key of the corresponding private key
+    #[clap(long, group = "public_key")]
+    public_key_path: Option<PathBuf>,
+    /// Path to json-serialized keypair
+    #[clap(long, short, group = "private_key", group = "public_key")]
+    keypair_path: Option<PathBuf>,
+    /// Unique id of blockchain
+    #[clap(long)]
+    chain_id: ChainId,
+    /// Path to genesis json file
+    #[clap(long, short)]
+    genesis_path: PathBuf,
+    /// Path to signed genesis output file
+    #[clap(long, short, group = "format")]
+    output_path: Option<PathBuf>,
+}
+
+impl GenesisSigningArgs {
+    fn get_private_key(&self) -> Result<PrivateKey> {
+        let private_key_bytes =
+            Self::get_key_bytes(&self.private_key_path, &self.private_key_string)?;
+        PrivateKey::from_bytes(self.algorithm.0, private_key_bytes.borrow()).wrap_err_with(|| {
+            eyre!(
+                "Failed to parse private key for algorithm `{}`",
+                self.algorithm
+            )
+        })
+    }
+
+    fn get_public_key(&self) -> Result<PublicKey> {
+        let public_key_bytes = Self::get_key_bytes(&self.public_key_path, &self.public_key_string)?;
+        PublicKey::from_bytes(self.algorithm.0, public_key_bytes.borrow()).wrap_err_with(|| {
+            eyre!(
+                "Failed to parse public key for algorithm `{}`",
+                self.algorithm
+            )
+        })
+    }
+
+    fn get_key_bytes<'a, P: AsRef<Path>>(
+        path: &Option<P>,
+        value: &'a Option<String>,
+    ) -> Result<Cow<'a, [u8]>, std::io::Error> {
+        match (path, value) {
+            (Some(path_buf), None) => Ok(Cow::Owned(fs::read(path_buf)?)),
+            (None, Some(hex)) => Ok(Cow::Borrowed(hex.as_bytes())),
+            _ => unreachable!("Clap group invariant"),
+        }
+    }
+
+    fn read_keypair<P: AsRef<Path>>(path: P) -> Result<KeyPair> {
+        let bytes = fs::read(path)?;
+        Ok(serde_json::from_slice(bytes.as_slice())?)
+    }
+}
+
+impl<T: Write> RunArgs<T> for GenesisSigningArgs {
+    fn run(self, writer: &mut BufWriter<T>) -> Outcome {
+        let key_pair = if let Some(path) = self.keypair_path {
+            Self::read_keypair(path)?
+        } else {
+            let (public_key, private_key) = (self.get_public_key()?, self.get_private_key()?);
+            KeyPair::new(public_key, private_key)?
+        };
+
+        let genesis_block = RawGenesisBlock::from_path(&self.genesis_path)?;
+        let encoded_genesis_network =
+            GenesisNetwork::new(genesis_block, &self.chain_id, &key_pair).encode();
+
+        if let Some(path) = self.output_path {
+            fs::write(&path, encoded_genesis_network)?;
+
+            writeln!(
+                writer,
+                "Genesis was successfully signed, encoded and written to `{}`",
+                path.display()
+            )?;
+        } else {
+            // TODO: ADD ?
```

Comment:
I think it would make sense to make `--output-path` a required arg.

Reply:
Left it as optional as we will output genesis in json to stdout as well